engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java

```diff
@@ -0,0 +1,178 @@
+package io.camunda.zeebe.process.test.engine;
+
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import io.camunda.zeebe.logstreams.log.LogStreamReader;
+import io.camunda.zeebe.logstreams.log.LoggedEvent;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.jupiter.api.Test;
+
+class EngineStateMonitorTest {
+
+  @Test
+  void testCallbackIsCalledWhenEngineIsAlreadyIdle()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // given
+    final InMemoryLogStorage logStorage = mock(InMemoryLogStorage.class);
+    final TestLogStreamReader logStreamReader = new TestLogStreamReader();
+    final EngineStateMonitor monitor = new EngineStateMonitor(logStorage, logStreamReader);
+
+    // when
+    final CompletableFuture<Void> callbackFuture = new CompletableFuture<>();
+    final Runnable callback = () -> callbackFuture.complete(null);
+    monitor.addOnIdleCallback(callback);
+
+    // then
+    callbackFuture.get(50L, TimeUnit.MILLISECONDS);
+    assertThat(callbackFuture).isCompleted();
+  }
+
+  @Test
+  void testCallbackIsCalledWhenIdleStateIsReached()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // given
+    final InMemoryLogStorage logStorage = mock(InMemoryLogStorage.class);
+    final TestLogStreamReader logStreamReader = new TestLogStreamReader();
+    final EngineStateMonitor monitor = new EngineStateMonitor(logStorage, logStreamReader);
+    changeToBusyState(monitor, logStreamReader);
```

Comment:
🤔 I wonder what would be clearer to read. We do the setup of the callbacks in _given_ and then in _when_ we change to busy state. Or we do it like it is now. But, personally, I find this hard to read. Because the when/then part is _when I register a callback, it is being called_. 

I like this test implementation, because it has an explicit call to _changeToBusyState_ I wonder whether we could also manufacture a _changeToIdleState_ method alongside and use it in the test case above.

Reply:
Ah, found the _changeToIdleState_ method.

Reply:
It's hard to structure the given/when/then like you suggested. This is because when we add a callback to the `EngineStateMonitor` it immediately checks if the engine is busy or idle, and then notifies the callbacks if necessary. To circumvent this we need to call `changeToBusyState` or `changeToIdleState` before we actually add the callback.

I will make some other changes to make it more readable and less confusing though.